### PR TITLE
feat: add prerender:route hook

### PR DESCRIPTION
**Pull request by TotomInc**
### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Add a new hook `prerender:route` that is called when a single page has been successfully prerendered.

This hook can be used externally to create specific logic tied to route generation, such as editing output of a route.

`generateRoute` now returns `{ contents, route, filePath }` instead of nothing, in order to:

- Access the saved file directly by its `filePath`.
- Read the `contents` that have been generated.

I can see more potential for SSG (fully static).

A quick example usage could be to dynamically download and replace static assets, stored on an API server, that you don't want to use in production (parse HTML content, download image, replace image path with the one downloaded locally).

Please, let me know if this change is not intended for nitro.

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/prerender.ts [hunk 1]
```diff
@@ -76,9 +78,16 @@ export async function prerender (nitro: Nitro) {
...
-      const error = await generateRoute(route).catch(err => err)
-      const end = Date.now()
-      nitro.logger.log(chalk.gray(`  ├─ ${route} (${end - start}ms) ${error ? `(${error})` : ''}`))
+      let end: number | null = null
+
+      try {
+        const { contents, filePath } = await generateRoute(route)
+        end = Date.now()
+        nitro.logger.log(chalk.gray(`  ├─ ${route} (${end - start}ms)`))
+        nitro.hooks.callHook('prerender:route', { route, contents, filePath })
```
Comment (pi0) [inline src/prerender.ts:87]: We need to await here
```suggestion
        await nitro.hooks.callHook('prerender:route', { route, contents, filePath })
```

File: src/prerender.ts [hunk 2]
```diff
@@ -76,9 +78,16 @@ export async function prerender (nitro: Nitro) {
...
+      } catch (error) {
+        nitro.logger.log(chalk.gray(`  ├─ ${route} (${end - start}ms) ${error ? `(${error})` : ''}`))
```
Comment (pi0) [inline src/prerender.ts:89]: I would also allow hook access to the errors rendered with error.

**Comment by pi0** [comment]
Thanks! This also resolves https://github.com/unjs/nitro/issues/88

File: src/prerender.ts (hunk 1 above)
Comment (TotomInc) [reply src/prerender.ts]: Thanks, added!

File: src/prerender.ts (hunk 2 above)
Comment (TotomInc) [reply src/prerender.ts:89]: Thanks for the feedback.

I've made the return response of the `prerender:routes` hook to be an object or an error. I also thought of introducing a hook to handle errors such as `prerender:route:error`. If you think it's cleaner to do this, I can mek the change.

Comment (pi0) [reply src/prerender.ts:89]: I was thinking, that by using a single hook, we can allow:

- Custom error (hooking to validate and modify context to extract an error being thrown)
- Clear error (hooking to allow still prerendering some pages with errors)

Comment (TotomInc) [reply src/prerender.ts:89]: Alright, I'm not sure I have understood correctly, but please let me know if my last changes seems good for you.

**Comment by pi0** [comment]
❤️

**Pull request merged**